pkg/kv/kvserver/replica_raftstorage.go

```diff
@@ -57,26 +57,69 @@ var snapshotIngestAsWriteThreshold = settings.RegisterByteSizeSetting(
 	}())
 
 // replicaRaftStorage implements the raft.Storage interface.
+//
+// All mutating calls to raft.RawNode require that r.mu is held. All read-only
+// calls to raft.RawNode require that r.mu is held at least for reads.
+//
+// All methods implementing raft.Storage are called from within, or on behalf of
+// a RawNode. When called from within RawNode, r.mu is held necessarily (and
+// maybe r.raftMu). When called from outside RawNode (on behalf of a RawNode
+// "snapshot"), the caller must hold r.raftMu and/or r.mu.
+//
+// RawNode has the in-memory "unstable" state which services most of its needs.
+// Most RawNode.Step updates are completed in memory, while only holding r.mu.
+//
+// RawNode falls back to reading from Storage when it does not have the needed
+// state in memory. For example, the leader may need to read log entries from
+// storage to construct a log append requests for a follower, or a follower may
+// need to interact with its storage upon receiving such a request to check
+// whether the appended log slice is consistent with raft rules.
+//
+// (1) There is a guarantee that everything RawNode reads from Storage has no
+// in-flight writes. Raft always reads state that it knows to be stable (meaning
```

Comment:
correction: no pending "append" type writes, like ones initiated by `RawNode`. There can be deletions/[truncations](https://github.com/cockroachdb/cockroach/pull/131041/files#r1767498347) though. We should be fine with that.